pkg/k8sutil/resource.go

```diff
@@ -105,25 +132,53 @@ func (r *GenericResourceReconciler) ReconcileResource(desired runtime.Object, de
 
 			currentResourceVersion, err := metaAccessor.ResourceVersion(current)
 			if err != nil {
-				return errors.Wrap(err, "failed to access resourceVersion from metadata")
+				return nil, errors.Wrap(err, "failed to access resourceVersion from metadata")
 			}
 			metaAccessor.SetResourceVersion(desired, currentResourceVersion)
 
 			var name string
 			if name, err = metaAccessor.Name(current); err != nil {
-				return errors.Wrap(err, "failed to access Name from metadata")
+				return nil, errors.Wrap(err, "failed to access Name from metadata")
 			}
 
 			log.V(1).Info("Updating resource",
 				"gvk", desired.GetObjectKind().GroupVersionKind(), "name", name)
 			if err := r.Client.Update(context.TODO(), desired); err != nil {
-				return emperror.WrapWith(err, "updating resource failed",
+				sErr, ok := err.(*apierrors.StatusError)
+				if ok && sErr.ErrStatus.Code == 422 && sErr.ErrStatus.Reason == metav1.StatusReasonInvalid {
+					if r.Logging.Spec.EnableRecreateWorkloadOnImmutableFieldChange {
+						r.Log.Error(err, "failed to update resource, trying to recreate")
+						err := r.Client.Delete(context.TODO(), current,
+							// wait until all dependent resources gets cleared up
+							runtimeClient.PropagationPolicy(metav1.DeletePropagationForeground),
+						)
+						if err != nil {
+							return nil, errors.Wrapf(err, "failed to delete resource %+v", current)
+						}
+						return &reconcile.Result{
+							Requeue:      true,
+							RequeueAfter: time.Second * 10,
+						}, nil
+					} else {
+						return nil, errors.New(
```

Comment:
Forgive my ignorance here, but will returning an error here cause reconciliation to stop? If so, is it safe to continue instead? If you run into this issue, you block all updates further on in the reconciliation loop, rather than updating what you can and reporting an error for what you cannot update.

Reply:
The other option would be to collect all the errors and return with all of them at once at the end of the loop. We could even write the error state into the status field for even better feedback on what is happening. What do you think?

Reply:
@worr @ahma Is it ok for you if I stick with the current method for now and create a follow up issue? I would like to do this in a broader context, not just for the reconciler, but for when we have issues with the model as well.

Reply:
Yeah, I think the current method definitely improves the current situation, regardless of changing how partial updates work. I'd be happy with a follow-up issue.